src/databricks/labs/remorph/helpers/validation.py

```diff
@@ -0,0 +1,95 @@
+# pylint: disable=wrong-import-order,ungrouped-imports, useless-suppression)
+import logging
+from io import StringIO
+
+from databricks.labs.lsql.backends import SqlBackend
+from databricks.sdk.errors.base import DatabricksError
+
+from databricks.labs.remorph.config import MorphConfig
+
+logger = logging.getLogger(__name__)
+
+
+class Validator:
+    """
+    The Validator class is used to validate SQL queries.
+    """
+
+    def __init__(self, sql_backend: SqlBackend):
+        self._sql_backend = sql_backend
+
+    def validate_format_result(self, config: MorphConfig, input_sql: str) -> tuple[str, str | None]:
+        """
+        Validates the SQL query and formats the result.
+
+        This function validates the SQL query based on the provided configuration. If the query is valid,
+        it appends a semicolon to the end of the query. If the query is not valid, it formats the error message.
+
+        Parameters:
+        - config (MorphConfig): The configuration for the validation.
+        - input_sql (str): The SQL query to be validated.
+
+        Returns:
+        - tuple: A tuple containing the result of the validation and the exception message (if any).
+        """
+        logger.debug(f"Validating query with catalog {config.catalog_name} and schema {config.schema_name}")
+        (is_valid, exception_msg) = self._query(self._sql_backend, input_sql)
+        if is_valid:
+            result = input_sql + "\n;\n"
+            exception_msg = None
+        else:
+            query = ""
+            if "[UNRESOLVED_ROUTINE]" in exception_msg:
+                query = input_sql
+            buffer = StringIO()
+            buffer.write("-------------- Exception Start-------------------\n")
+            buffer.write("/* \n")
+            buffer.write(exception_msg)
+            buffer.write("\n */ \n")
+            buffer.write(query)
+            buffer.write("\n ---------------Exception End --------------------\n")
+
+            result = buffer.getvalue()
+
+        return result, exception_msg
+
+    def _query(self, sql_backend: SqlBackend, query: str) -> tuple[bool, str | None]:
+        """
+        Validate a given SQL query using the provided SQL backend
+
+        Parameters:
+        - query (str): The SQL query to be validated.
+        - sql_backend (SqlBackend): The SQL backend to be used for validation.
+
+        Returns:
+        - tuple: A tuple containing a boolean indicating whether the query is valid or not,
+        and a string containing a success message or an exception message.
+        """
+        # When variables is mentioned Explain fails we need way to replace them before explain is executed.
+        explain_query = f'EXPLAIN {query.replace("${", "`{").replace("}", "}`").replace("``", "`")}'
+        try:
+            rows = list(sql_backend.fetch(explain_query))
+            if not rows:
+                return False, "No results returned from explain query."
```

Comment:
do we need this check given explain will always return rows or error out.

Reply:
It is just for safeguarding against erroneous result.